mne/preprocessing/_csd.py

```diff
@@ -0,0 +1,176 @@
+# Copyright 2017 Federico Raimondo <federaimondo@gmail.com> and
+#                Denis A. Engemann <dengemann@gmail.com>
+#
+#
+# The original CSD Toolbox can be find at
+# http://psychophysiology.cpmc.columbia.edu/Software/CSDtoolbox/
+
+# Authors: Denis A. Engeman <denis.engemann@gmail.com>
+#          Alex Rockhill <aprockhill206@gmail.com>
+#
+# License: Relicensed under BSD (3-clause) and adapted with
+#          permission from authors of original GPL code
+
+import numpy as np
+
+from scipy.linalg import inv
+
+from mne import pick_types, pick_info
+from mne.utils import _validate_type, _check_ch_locs, _ensure_int
+from mne.io import BaseRaw
+from mne.io.constants import FIFF
+from mne.epochs import BaseEpochs
+from mne.evoked import Evoked
+from mne.bem import fit_sphere_to_headshape
+from mne.channels.interpolation import _calc_g, _calc_h
+
+
+def _prepare_G(G, lambda2):
+    G.flat[::len(G) + 1] += lambda2
+    # compute the CSD
+    Gi = inv(G)
+
+    TC = Gi.sum(0)
+    sgi = np.sum(TC)  # compute sum total
+
+    return Gi, TC, sgi
+
+
+def _compute_csd(G_precomputed, H, radius):
+    """Compute the CSD."""
+    n_channels = H.shape[0]
+    data = np.eye(n_channels)
+    mu = data.mean(0)
+    Z = data - mu
+
+    Gi, TC, sgi = G_precomputed
+
+    Cp2 = np.dot(Gi, Z)
+    c02 = np.sum(Cp2, axis=0) / sgi
+    C2 = Cp2 - np.dot(TC[:, None], c02[None, :])
+    X = np.dot(C2.T, H).T / radius ** 2
+    return X
+
+
+def compute_current_source_density(inst, lambda2=1e-5, stiffness=4,
+                                   n_legendre_terms=50, sphere='auto',
+                                   copy=True):
+    """Get the current source density (CSD) transformation.
+
+    Transformation based on spherical spline surface Laplacian.
+
+    .. note:: This function applies an average reference to the data.
+              Do not transform CSD data to source space.
+
+    Parameters
+    ----------
+    inst : instance of Raw, Epochs or Evoked
+        The data to be transformed.
+    lambda2 : float
+        Regularization parameter, produces smoothnes. Defaults to 1e-5.
+    stiffness : float
+        Stiffness of the spline. Also referred to as `m`.
+    n_legendre_terms : int
+        Number of Legendre terms to evaluate.
+    sphere : array-like, shape (4,)
+        The sphere, head-model of the form (x, y, z, r) where x, y, z
+        is the center of the sphere and r is the radius in meters.
+    copy : bool
+        Whether to overwrite instance data or create a copy.
+
+    Returns
+    -------
+    inst_csd : instance of Epochs or Evoked
+        The transformed data. Output type will match input type.
+    """
+    _validate_type(inst, (BaseEpochs, BaseRaw, Evoked), 'inst')
+
+    if inst.info['custom_ref_applied'] == -1:
+        raise ValueError('CSD already applied, should not be reapplied')
+
+    inst = inst.copy() if copy else inst
+
+    picks = pick_types(inst.info, meg=False, eeg=True, exclude='bads')
```

Comment:
This should guarantee that all picked channels are EEG, so no need to check COIL_EEG below.

Also it would be good if, after this operation, there were no bad EEG channels, otherwise we end up in a weird state with some EEG channels (all of which are bad) and some EEG-CSD channels (all of which are good). So we should either:

1. Raise an error if any EEG channels are bad they should be interpolated
2. Automatically interpolate the bad EEG channels before doing the CSD

I'd vote for option (2).

Reply:
Hmmm, after thinking about it I could see two cases:

1) Someone wants the bad channels interpolated and used in the csd Laplacian.

If this were the case I think it would be best to let them do that before calling compute_current_source_density

2) Someone wants to completely ignore the bad channels 

If this were the case, the bad channels could dropped from the data and excluded from the info. This would have to be added to the csd function. 

@larsoner what do you think?

Reply:
The simplest solution is just to raise an error if there are bad EEG channels, and tell them to either interpolate then our drop them, rather than us forcing that choice either way

Reply:
Sounds good